network/src/internal/message_handler.rs

```diff
@@ -309,58 +332,79 @@ impl Server {
     /// Add all new/updated addresses to our gossiped.
     /// The connection handler will be responsible for sending out handshake requests to them.
     async fn receive_peers(&mut self, message: Peers, channel: Arc<Channel>) -> Result<(), ServerError> {
-        let peer_book = &mut self.context.peer_book.write().await;
-        for (addr, time) in message.addresses.iter() {
-            if &*self.context.local_address.read().await == addr {
+        // If we received a message, but aren't connected to the peer,
+        // inform the peer book that we found a peer.
+        // The peer book will determine if we have seen the peer before,
+        // and include the peer if it is new.
+        let mut peer_book = self.context.peer_book.write().await;
+        if !peer_book.is_connected(&channel.address) {
+            peer_book.found_peer(&channel.address);
+        }
+
+        // Process all of the peers sent in the message,
+        // by informing the peer book of that we found peers.
+        let local_address = *self.context.local_address.read().await;
+
+        for (peer_address, _) in message.addresses.iter() {
+            // Skip if the peer address is the node's local address.
+            let is_zero_address = match "0.0.0.0".to_string().parse::<IpAddr>() {
+                Ok(zero_ip) => (*peer_address).ip() == zero_ip,
+                _ => false,
+            };
+            if *peer_address == local_address || is_zero_address {
                 continue;
-            } else if peer_book.connected_contains(addr) {
-                peer_book.update_connected(addr.clone(), time.clone());
-            } else {
-                peer_book.update_gossiped(addr.clone(), time.clone());
+            }
+            // Inform the peer book that we found a peer.
+            // The peer book will determine if we have seen the peer before,
+            // and include the peer if it is new.
+            else if !peer_book.is_connected(&channel.address) {
+                peer_book.found_peer(&channel.address);
```

Comment:
this has already been done at lines 340-342; also, it doesn't seem to benefit from being done in a loop - `channel.address` and `peer_book` don't seem to change in the meantime; removing this would also allow the `peer_book` lock to be dropped earlier, e.g. at like 343